Feature/search container and modes

CorreyL opened the pull request:
## Note

This PR temporarily merges to `feature/text-search` to keep the diff managable. I will target `master` when `feature/text-search` is merged, per this pending PR: https://github.com/PDFTron/webviewer-custom-ui/pull/2

## Preamble

This PR concerns moving all the search related logic and interface elements into a `SearchContainer` component to ensure we follow the [Separation of Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) design pattern.

This PR also adds functionality to allow the user to toggle `Case sensitive` and `Whole word` searching modes.

![search-prev-next](https://user-images.githubusercontent.com/16601729/85079986-ef798100-b17c-11ea-9f97-e490db9760ed.gif)

CorreyL committed:
Add boilerplate for a SearchContainer component

CorreyL committed:
Render SearchContainer and add button to toggle

CorreyL committed:
Add logic to edit the styles of HTML elements

Add commented out code that would change the `width` attribute of the
`div#scroll-view` and the returned `div` from `SearchContainer`, such
that the `SearchContainer` component renders to the right of the
`div#scroll-view` element

Comment out the logic for this for now, as it is not behaving as
expected, due to the `div#scroll-view` element always being treated
as a `block` level element, rather than `inline`, thus causing the
`SearchContainer` component to always be rendered below the
`div#scroll-view` element

CorreyL committed:
Move search logic & HTML into SearchContainer

Move all the search logic and HTML elements implemented in the component
`App` to the `SearchContainer` component, ensuring the
`separation of concerns` design pattern is followed, making the code
cleaner to read

CorreyL committed:
Replace `Open` string in button with `Search` icon

CorreyL committed:
Clear previous search results on new search

Add boolean param `clearSearchTermValue` for a guard clause against
modifying `searchTerm.current.value`, so that `performSearch` can safely
invoke `clearSearchResults`. Give `clearSearchTermValue` a default value
of `true` to ensure previously committed invocations of
`clearSearchResults` continue to behave as expected

CorreyL committed:
Add case sensitivity and whole word toggling

- Define `searchModes` as a state value held in an object by adding a
  `useEffect` watcher on `docViewer`, and extracting the `searchMode`
  values of interest when they are available (i.e. when `docViewer` is
  fully instantiated)
- Store toggled search modes in `toggledSearchModes`, and utilize them
  by performing the bit-wise OR operation by iterating through
  `toggledSearchModes` with a `.reduce` call. Set `ePageStop` and
  `eHighlight` as the initial value of the `reduce` call, since they
  were previously always toggled on in previous commits
- Define `toggleSearchMode` and invoke it inside of
  `toggleCaseSensitive` and `toggleWholeWorld` to ensure the code sticks
  to the DRY (don't repeat yourself) principle
- Add HTML `input` elements with `checkbox` types to allow the user to
  toggle the `Case sensitive` and `Whole word` modes

CorreyL committed:
Remove redundant input closing tag

Make `input` a single HTML tag for the sake of brevity

CorreyL committed:
Add clause to not exceed length of the array

Originally implemented in 991299e

CorreyL committed:
Add doc-string for performSearch

andreysaf reviewed (approved):
Looks great! I will leave merging up to you since you might want to retarget as necessary.

CorreyL committed:
Update state inside of click handlers

For the sake of code simplicity, by reducing the number of defined
functions in the body of the component

As originally implemented in: 5e8280a

CorreyL committed:
Merge branch master

Into `feature/search-container-and-modes` to resolve merge conflict

CorreyL commented:
> I will leave merging up to you since you might want to retarget as necessary.

Looks like GitHub took care of the re-target automatically for me, which I had no idea it would do that. 👀 🎉
